File: FundComponents/ImpactAgricultureComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components.ImpactAgriculture
{

    public interface IImpactAgricultureState
    {
        IParameter1Dimensional<Region, double> gdp90 { get; }
        IParameter2Dimensional<Timestep, Region, double> income { get; }
        IParameter1Dimensional<Region, double> pop90 { get; }
        IParameter2Dimensional<Timestep, Region, double> population { get; }

        IVariable2Dimensional<Timestep, Region, double> agrish { get; }
        IParameter1Dimensional<Region, double> agrish0 { get; }
        Double agel { get; }

        IVariable2Dimensional<Timestep, Region, double> agrate { get; }
        IVariable2Dimensional<Timestep, Region, double> aglevel { get; }
        IVariable2Dimensional<Timestep, Region, double> agco2 { get; }
        IVariable2Dimensional<Timestep, Region, double> agcost { get; }

        IParameter1Dimensional<Region, double> agrbm { get; }
        IParameter1Dimensional<Region, double> agtime { get; }
        Double agnl { get; }

        IParameter1Dimensional<Region, double> aglparl { get; }
        IParameter1Dimensional<Region, double> aglparq { get; }

        IParameter1Dimensional<Region, double> agcbm { get; }
        Double co2pre { get; }

        IParameter2Dimensional<Timestep, Region, double> temp { get; }
        IParameter1Dimensional<Timestep, double> acco2 { get; }
    }

    public class ImpactAgricultureComponent
    {
        public const double DBsT = 0.04;     // base case yearly warming

        public void Run(Clock clock, IImpactAgricultureState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            if (clock.IsFirstTimestep)
            {
                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.agrate[t, r] = s.agrbm[r] * Math.Pow(0.005 / DBsT, s.agnl) * s.agtime[r];
                }
            }
            else
            {
                foreach (var r in dimensions.GetValues<Region>())
                {
                    Double ypc = s.income[t, r] / s.population[t, r] * 1000.0;
                    Double ypc90 = s.gdp90[r] / s.pop90[r] * 1000.0;

                    s.agrish[t, r] = s.agrish0[r] * Math.Pow(ypc / ypc90, -s.agel);
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    var dtemp = Math.Abs(s.temp[t, r] - s.temp[t - 1, r]);

                    if (double.IsNaN(Math.Pow(dtemp / 0.04, s.agnl)))
                        s.agrate[t, r] = 0.0;
                    else
                        s.agrate[t, r] = s.agrbm[r] * Math.Pow(dtemp / 0.04, s.agnl) + (1.0 - 1.0 / s.agtime[r]) * s.agrate[t - 1, r];
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.aglevel[t, r] = s.aglparl[r] * s.temp[t, r] + s.aglparq[r] * Math.Pow(s.temp[t, r], 2.0);
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.agco2[t, r] = s.agcbm[r] / Math.Log(2.0) * Math.Log(s.acco2[t - 1] / s.co2pre);
                }

                foreach (var r in dimensions.GetValues<Region>())
                {
                    s.agcost[t, r] = Math.Min(1.0, s.agrate[t, r] + s.aglevel[t, r] + s.agco2[t, r]) * s.agrish[t, r] * s.income[t, r];
                }
            }
        }
    }
}

File: src/components/ImpactAgricultureComponent.jl
@defcomp impactagriculture begin
    regions = Index()

    gdp90 = Parameter(index=[regions])
    income = Parameter(index=[time,regions])
    pop90 = Parameter(index=[regions])
    population = Parameter(index=[time,regions])

    agrish = Variable(index=[time,regions])
    agrish0 = Parameter(index=[regions])
    agel = Parameter(default=0.31)

    agrate = Variable(index=[time,regions])
    aglevel = Variable(index=[time,regions])
    agco2 = Variable(index=[time,regions])
    agcost = Variable(index=[time,regions])

    agrbm = Parameter(index=[regions])
    agtime = Parameter(index=[regions])
    agnl = Parameter(default=2)

    aglparl = Parameter(index=[regions])
    aglparq = Parameter(index=[regions])

    agcbm = Parameter(index=[regions])
    co2pre = Parameter()

    temp = Parameter(index=[time,regions])
    acco2 = Parameter(index=[time])

    function run_timestep(p, v, d, t)
        
        DBsT = 0.04     # base case yearly warming
        
        if is_first(t)
            for r in d.regions
                v.agrate[t, r] = p.agrbm[r] * (0.005 / DBsT)^p.agnl * p.agtime[r]
            end
        else
            for r in d.regions
                ypc = p.income[t, r] / p.population[t, r] * 1000.0
                ypc90 = p.gdp90[r] / p.pop90[r] * 1000.0

                v.agrish[t, r] = p.agrish0[r] * (ypc / ypc90)^(-p.agel)
            end

            for r in d.regions
                dtemp = abs(p.temp[t, r] - p.temp[t - 1, r])

                if isnan((dtemp / 0.04)^p.agnl)
                    v.agrate[t, r] = 0.0
                else
                    v.agrate[t, r] = p.agrbm[r] * (dtemp / 0.04)^p.agnl + (1.0 - 1.0 / p.agtime[r]) * v.agrate[t - 1, r]
                end
            end

            for r in d.regions
                v.aglevel[t, r] = p.aglparl[r] * p.temp[t, r] + p.aglparq[r] * p.temp[t, r]^2.0
            end

            for r in d.regions
                v.agco2[t, r] = p.agcbm[r] / log(2.0) * log(p.acco2[t - 1] / p.co2pre)
            end

            for r in d.regions
                v.agcost[t, r] = min(1.0, v.agrate[t, r] + v.aglevel[t, r] + v.agco2[t, r]) * v.agrish[t, r] * p.income[t, r]
            end
        end
    end
end
